src/backend/cuda/nvrtc/cache.cpp

```diff
@@ -149,6 +157,107 @@ void Kernel::getScalar(T &out, const char *name) {
 template void Kernel::setScalar<int>(const char *, int);
 template void Kernel::getScalar<int>(int &, const char *);
 
+#if !defined(OS_WIN)
+string getHomeDirectory() {
+    string home = getEnvVar("XDG_CONFIG_HOME");
+    if (!home.empty()) return home;
+
+    home = getEnvVar("HOME");
+    if (!home.empty()) return home;
+
+    home = getpwuid(getuid())->pw_dir;
+}
+#else
+string getTemporaryDirectory() {
+    DWORD bufSize = 261;  // limit according to GetTempPath documentation
+    string retVal;
+    retVal.resize(bufSize);
+    bufSize = GetTempPathA(bufSize, &retVal[0]);
+    retVal.resize(bufSize);
+    return retVal;
+}
+#endif
+
+bool folderExists(const string &path) {
+#if !defined(OS_WIN)
+    struct stat status;
+    return stat(path.c_str(), &status) == 0 && (status.st_mode & S_IFDIR) != 0;
+#else
+    struct _stat status;
+    return _stat(path.c_str(), &status) == 0 && (status.st_mode & S_IFDIR) != 0;
+#endif
+}
+
+bool createFolder(const string &path) {
+#if !defined(OS_WIN)
+    return mkdir(path.c_str(), 0777) == 0;
+#else
+    return CreateDirectoryA(path.c_str(), NULL) != 0;
+#endif
+}
+
+bool removeFile(const string &path) {
+#if !defined(OS_WIN)
+    return unlink(path.c_str()) == 0;
+#else
+    return DeleteFileA(path.c_str()) != 0;
+#endif
+}
+
+bool isDirectoryWritable(const string &path) {
+    if (!folderExists(path) && !createFolder(path)) return false;
+
+    const string testPath = path + AF_PATH_SEPARATOR + "test";
+    if (!std::ofstream(testPath).is_open()) return false;
+    removeFile(testPath);
+    
+    return true;
+}
+
+const string &getKernelCacheDirectory() {
+    thread_local std::once_flag flag;
+    thread_local string cacheDirectory;
```

Comment:
My understanding is that a file cannot be opened twice with write permission. This means that two threads cannot write the same kernel file at the same time; one of the threads will fail to open the file for writing. By default, ``std`` streams fail silently so this will simply do nothing, both threads will continue normally, and the file will be written on disk correctly.

The other risk is if one thread is writing the file while another thread tries to read that same file. This can happen, although because the byte size of the CUBIN is included in the file itself, this acts as some crude form of checksum, and there will be a read error (not enough bytes to read from file). Currently, this will make ``loadKernel()`` throw an exception. We can make it so any exception is caught and an empty ``Kernel{nullptr, nullptr}`` is returned instead, so the thread will fallback to building the kernel again.

Reply:
I didn't know that std stream are thread safe. From what I found in the following links
- [Only iostream is thread safe, even then the order of characters/words in the output cannot be deterministically identified. ](https://docs.microsoft.com/en-us/cpp/standard-library/thread-safety-in-the-cpp-standard-library?redirectedfrom=MSDN&view=vs-2019)
- [So, if your platform's C library is threadsafe, then your fstream I/O operations will be threadsafe at the lowest level. For higher-level operations, such as manipulating the data contained in the stream formatting classes (e.g., setting up callbacks inside an std::ofstream), you need to guard such accesses like any other critical shared resource. ](https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_concurrency.html)

There is no guarantee that the behavior is well defined with file streams in C++. They have to be guarded to be safe in a multi-threaded context. A similar discussion on stackoverflow can be found here https://stackoverflow.com/a/15513639/1985476

@umar456 what are you thoughts ?

Reply:
I think the links above are mostly about accessing the same stream object from multiple threads, which is not what we are doing here. The same file is accessed by different streams (which will have a different underlying file handles too), so the thread-safety is down to the file system and whether it will allow a file being open for concurrent write.

I just did a test on my Linux and Windows machines with running two instances of the following program 
```c++
#include <fstream>
#include <iostream>
#include <string>

int main() {
    // Open the file
    std::ofstream out("test.txt");
    out.exceptions(std::ios::failbit | std::ios::badbit);

    // Wait for user signal to write to it
    int c;
    std::cin >> c;
    std::cout << out.is_open() << std::endl;

    // Write and close
    out << c;
    out.close();

    return 0;
}
```

On both OSes, the two program instances were able to open ``test.txt`` and write to it concurrently. So not thread-safe. I expected this wouldn't be possible...

That being said, the two threads would be writing exactly the same content to the file, so I don't expect this would be an issue in practice. If we want to make this more thread-safe, we can wrap any operation on cached kernels files by a mutex. That would solve multi-threaded access within one process, but not data races across multiple processes.

Reply:
C++ streams use ``fopen()`` [under the hood](https://en.cppreference.com/w/cpp/io/basic_filebuf/open).

According to the [C standard](https://wiki.sei.cmu.edu/confluence/display/c/FIO24-C.+Do+not+open+a+file+that+is+already+open), whether calling ``fopen()`` twice on the same file is allowed or not is implementation-defined. For those OSes that don't allow it, we're fine, because the file will just refuse to open twice and only one write will occur. From the above however, it looks like common OSes do allow it.

According to [this SO question](https://stackoverflow.com/questions/11723397/opening-a-file-using-fopen-with-same-flag-in-c), if the files are small and if we don't open the files in "append" mode, there is no risk of interleaving data in the output file, as the last thread that calls ``close()`` will flush its buffer and overwrite the content written by other threads. Although here I'm sure we're also in implementation-defined land (up to which file size will the OS use such buffering?).

I've open an [SO question](https://stackoverflow.com/questions/61339944/is-it-safe-for-two-threads-to-write-identical-content-to-the-same-file) to ask for further inputs.

Reply:
It's been proposed on SO that we should write the kernel to a temporary file, and rename it to the final cached file once the writing is complete. This prevents concurrent writes to the same file, and incidentally, it also ensures that whenever a thread finds a cached kernel file, the file is complete. If you agree, I will implement this.

Reply:
I find it reasonable, let's wait for @umar456 thoughts on the idea ? I will ping him tonight.

Thanks for checking out the potential scenarios. This is going to be a great asset for kernel caching.

Reply:
FYI I've implemented the temporary file + rename and pushed the changes.